Java-Interfaces.md

```diff
@@ -0,0 +1,219 @@
+What can you do with this feature? Polymorphism! You can use only interfaces to refer to object instances!
+
+```java
+class Truck implements Vehicle {
+	public void start() {
+		System.out.println("starting truck engine...");
+	}
+	public void stop() {
+		System.out.println("stopping truck engine...");
+	}
+}
+
+class Starter {
+	// static method, can be called without instantiating the class
+	public static void startEngine(Vehicle vehicle) {
+		vehicle.start();
+	}
+}
+
+Vehicle tesla = new Car();
+Vehicle tata = new Truck();
+
+Starter.startEngine(tesla); // starting engine ...
+Starter.startEngine(tata); // starting truck engine ...
+
+```
+
+## But how about multiple interfaces?
+
+Yes, you can implement multiple Interfaces in a single class. But do not forget to implement _all_ of the methods of all the Interfaces, otherwise compilation will fail!
+
+```java
+public interface GPS {
+	public LatLng getCoordinates();
+}
+
+public interface Radio {
+	public void startRadio();
+	public void stopRadio();
+}
+
+public class Smartphone implements GPS,Radio {
+	public LatLng getCoordinates() {
+		// return some coordinates
+	}
+	public void startRadio() {
+	  // start Radio
+	}
+	public void stopRadio() {
+		// stop Radio
+	}
+}
+```
+
+## Some features of Interfaces
+
+* You can place variables within an Interface, although it wont be a sensible decision as Classes are not bound to have the same variable. In short, avoid placing variables!
+* All variables and methods in an Interface are public, even if you leave out the `public` keyword.
+* An Interface cannot specify the implementation of a particular method. Its upto the Classes to do it. Although there has been a recent exception (see below).
+* If a Class implements multiple Interfaces, then there is a remote chance of method signature overlap. Since Java does not allow multiple methods of the exact same signature, this can lead to problems. See [this question](http://stackoverflow.com/questions/2598009/method-name-collision-in-interface-implementation-java) for more info!
+
+## Interface Default Methods
+
+Before Java 8, we had no way to direct an Interface to have a particular method implementation. This lead to lot of confusion and code breaks if an Interface definition is suddenly changed.
+
+Suppose you wrote an open source library, which contains an Interface. Say your clients, i.e. practically all developers around the world, are using it heavily and are happy. Now you had to upgrade the library by adding a new method definition to the Interface to support a new feature. But that would break _all_ builds since all Classes implementing that Interface has to change now! What a catastrophe!
+
+Thankfully, Java 8 now provides us `default` methods of Interfaces. A `default` method _can_ contain its own implementation _directly_ within the Interface! So, if a Class does not implement a default method, the compiler will take the implementation mentioned within the Interface. Nice, isnt it?
```

Comment:
Mention how this solves the earlier problem of broken build, when you update one method in a top level interface.